code/datums/components/autofire.dm

```diff
@@ -0,0 +1,124 @@
+/datum/component/automatedfire/autofire
+	///The current fire mode of the shooter
+	var/fire_mode
+	///Delay between two shot when in full auto
+	var/auto_fire_shot_delay
+	///Delay between two burst shots
+	var/burstfire_shot_delay
+	///How many bullets are fired in burst mode
+	var/burst_shots_to_fire
+	///Count the shots fired when bursting
+	var/shots_fired = 0
+	///If the shooter is currently shooting
+	var/shooting = FALSE
+	///If TRUE, the shooter will reset its references at the end of the burst
+	var/have_to_reset = FALSE
+	///If we are in a burst
+	var/bursting = FALSE
+
+/datum/component/automatedfire/autofire/Initialize(_auto_fire_shot_delay = 0.3 SECONDS, _burstfire_shot_delay, _burst_shots_to_fire = 3, _fire_mode = GUN_FIREMODE_SEMIAUTO)
+	. = ..()
+	if(!isatom(parent))
+		return COMPONENT_INCOMPATIBLE
```

Comment:
Is this really usefull?

Reply:
should be atom/movable i think unless you want turfs that can shoot.